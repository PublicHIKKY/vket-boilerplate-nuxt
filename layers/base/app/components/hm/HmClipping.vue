<template>
  <div class="hm-clipping">
    <div class="cropper-container">
      <Cropper
        ref="cropper"
        class="cropper"
        :src="src"
        :auto-zoom="autoZoom"
        :stencil-size="{
          width: width,
          height: height,
        }"
        v-bind="cropperOptions"
        default-boundaries="fit"
        :image-restriction="imageRestriction"
        :style="forceStyle"
        @change="onChange"
      />
    </div>
    <template v-if="src">
      <HaBaseButton
        class="button"
        @click="clip"
      >
        <!-- {{ i18n.t('label') }} -->
        切り抜く
      </HaBaseButton>
    </template>
  </div>
</template>

<script lang="ts" setup>
import { Cropper } from 'vue-advanced-cropper'
import 'vue-advanced-cropper/dist/style.css'

type Props = {
  src: string
  width?: number
  height?: number
  cropperAreaHeight?: number
  doResize?: boolean
  stencil?: string
  imageRestriction?: 'fill-area' | 'fit-area' | 'stencil' | 'none'
  autoZoom?: boolean
  ext?: string
}

const props = withDefaults(defineProps<Props>(), {
  width: 256,
  height: 256,
  cropperAreaHeight: undefined,
  doResize: true,
  stencil: 'RectangleStencil',
  imageRestriction: 'stencil',
  autoZoom: false,
  ext: 'jpeg',
})

type Emits = {
  (e: 'clipped', image: File[]): void
}
const emit = defineEmits<Emits>()

const croppedImage = ref<File[]>([])

const forceStyle = computed(() => {
  if (props.cropperAreaHeight === undefined) return {}
  const cropperAreaHeight = props.cropperAreaHeight || props.height * 1.4
  return {
    height: `${cropperAreaHeight}px`,
  }
})
const cropperOptions = computed(() => {
  return props.doResize
    ? {
        canvas: {
          width: props.width,
          height: props.height,
        },
      }
    : {}
})

const onChange = ({ canvas }: { canvas: HTMLCanvasElement }) => {
  /*
   * this.coordinates = coordinates
   * note: canvas to DataURI
   * this.croppedImage = canvas.toDataURL(`image/${this.ext}`)
   */
  const data = canvas.toDataURL(`image/${props.ext}`)
  // note: DataURL to File
  const bytes = atob(data.split(',')[1] ?? raiseError('Invalid bytes'))
  const mime
    = data.split(',')[0]?.split(':')[1]?.split(';')[0]
      ?? raiseError('Invalid mime')
  const name = `tmp-${new Date().getTime()}.${mime.split('/')[1]}`
  const writer = new Uint8Array(new ArrayBuffer(bytes.length))
  for (let i = 0; i < bytes.length; i++) {
    writer[i] = bytes.charCodeAt(i)
  }
  const file = new File([writer.buffer], name, { type: mime })
  croppedImage.value[0] = file
}

const clip = () => {
  emit('clipped', croppedImage.value)
}
</script>

<style lang="scss" scoped>
@use '@/assets/styles/variables' as v;
@use '@/assets/styles/mixins' as m;

.hm-clipping {
  height: 100%;
  width: 100%;

  > .cropper-container {
    background: #000;
    height: calc(100% - 70px);
    margin-bottom: 20px;
    min-height: 300px;
    padding: v.space(2);
  }

  > .cropper-container > .cropper {
    background: #000;
    height: 100%;
  }

  > .button {
    background-color: v.$primary-button-default-color;
    color: v.$white;
    padding: v.$space-small;
    width: 100%;

    :hover {
      background-color: v.$primary-button-active-color;
    }
  }
}
</style>
