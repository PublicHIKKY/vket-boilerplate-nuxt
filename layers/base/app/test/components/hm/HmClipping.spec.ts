import { mount, shallowMount } from '@vue/test-utils'
import HmClipping from '#base/app/components/hm/HmClipping.vue'

test('ref component', () => {
  expect(HmClipping).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HmClipping, {
    props: {
      src: '',
    },
  })
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

test('props', () => {
  const wrapper = mount(HmClipping, {
    props: {
      src: '',
      width: 256,
      height: 256,
      cropperAreaHeight: 0,
      doResize: true,
      stencil: 'RectangleStencil',
      imageRestriction: 'stencil',
      autoZoom: false,
      ext: 'jpeg',
    },
  })
  expect(wrapper.props('src')).toStrictEqual('')
  expect(wrapper.props('width')).toStrictEqual(256)
  expect(wrapper.props('height')).toStrictEqual(256)
  expect(wrapper.props('cropperAreaHeight')).toStrictEqual(0)
  expect(wrapper.props('doResize')).toStrictEqual(true)
  expect(wrapper.props('stencil')).toStrictEqual('RectangleStencil')
  expect(wrapper.props('imageRestriction')).toStrictEqual('stencil')
  expect(wrapper.props('autoZoom')).toStrictEqual(false)
  expect(wrapper.props('ext')).toStrictEqual('jpeg')
})

describe('events', () => {
  it(':button click to emit clipped', async () => {
    // NOTE: mountしてcomponentを展開すると、「Error: connect ECONNREFUSED」になる
    const wrapper = shallowMount(HmClipping, {
      props: {
        src: '/dummy',
        width: 256,
        height: 256,
        cropperAreaHeight: 0,
        doResize: true,
        stencil: 'RectangleStencil',
        imageRestriction: 'stencil',
        autoZoom: false,
        ext: 'jpeg',
      },
    })
    await wrapper.find('ha-base-button-stub').trigger('click')
    expect(wrapper.emitted()).toHaveProperty('clipped')
    expect(wrapper.emitted()['clipped']).toHaveLength(1)
    expect(wrapper.emitted()['clipped']).toEqual([[[]]])
  })

  // TODO: Cropperのchangeのテスト
  it(':Cropper change', () => {
    // NOTE: mountしてcomponentを展開すると、「Error: connect ECONNREFUSED」になる
    const _wrapper = shallowMount(HmClipping, {
      props: {
        src: '/dummy',
        width: 256,
        height: 256,
        cropperAreaHeight: 0,
        doResize: true,
        stencil: 'RectangleStencil',
        imageRestriction: 'stencil',
        autoZoom: false,
        ext: 'jpeg',
      },
    })
    /*
     * ERROR: 発火はできるが、canvas.toDataURLが読めず、vi.importActualにてcanvas.toDataURLのみを偽装してもエラーとなったのでコメントアウトする
     * await wrapper.find('cropper-stub').trigger('change')
     */
  })
})
