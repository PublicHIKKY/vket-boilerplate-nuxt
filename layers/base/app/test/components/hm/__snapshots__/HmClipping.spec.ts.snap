// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `
"<div data-v-7cbd279a="" class="hm-clipping">
  <div data-v-7cbd279a="" class="cropper-container">
    <div data-v-7cbd279a="" class="vue-advanced-cropper cropper">
      <div class="vue-advanced-cropper__stretcher"></div>
      <div class="vue-advanced-cropper__boundaries" style="width: auto; height: auto; transition: opacity 300ms; pointer-events: none; opacity: 0;">
        <div class="vue-advanced-cropper__cropper-wrapper">
          <div class="vue-advanced-cropper__background" style="width: auto; height: auto; transition: opacity 300ms; pointer-events: none; opacity: 0;"></div>
          <div class="vue-advanced-cropper__image-wrapper"><img class="vue-advanced-cropper__image" style="height: 0px; left: 0px; top: 0px; transform: translate(NaNpx, NaNpx) rotate(0deg)  scaleX(NaN)  scaleY(NaN) ;"></div>
          <div class="vue-advanced-cropper__foreground" style="width: auto; height: auto; transition: opacity 300ms; pointer-events: none; opacity: 0;"></div>
          <div class="vue-rectangle-stencil vue-rectangle-stencil--movable" style="width: 0px; height: 0px; transform: translate(0px, 0px); display: none;">
            <div class="vue-bounding-box vue-rectangle-stencil__bounding-box">
              <div>
                <div class="vue-preview vue-preview--fill vue-rectangle-stencil__preview">
                  <div class="vue-preview__wrapper" style="width: 0px; height: 0px; left: calc(50% - 0px); top: calc(50% - 0px);"><img class="vue-preview__image" style="width: 0px; height: 0px; left: 0px; top: 0px; transform: translate(
				NaNpx,NaNpx)  rotate(0deg)  scaleX(NaN)  scaleY(NaN) ; display: none;"></div>
                </div>
              </div>
              <div>
                <div class="vue-line-wrapper vue-line-wrapper--east vue-simple-line-wrapper vue-simple-line-wrapper--east">
                  <div class="vue-simple-line vue-simple-line--east"></div>
                </div>
                <div class="vue-line-wrapper vue-line-wrapper--west vue-simple-line-wrapper vue-simple-line-wrapper--west">
                  <div class="vue-simple-line vue-simple-line--west"></div>
                </div>
                <div class="vue-line-wrapper vue-line-wrapper--south vue-simple-line-wrapper vue-simple-line-wrapper--south">
                  <div class="vue-simple-line vue-simple-line--south"></div>
                </div>
                <div class="vue-line-wrapper vue-line-wrapper--north vue-simple-line-wrapper vue-simple-line-wrapper--north">
                  <div class="vue-simple-line vue-simple-line--north"></div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--east-south">
                <div class="vue-handler-wrapper vue-handler-wrapper--east-south vue-simple-handler-wrapper vue-simple-handler-wrapper--east vue-simple-handler-wrapper--south vue-simple-handler-wrapper--east-south vue-bounding-box__handler vue-bounding-box__handler--east-south">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--east vue-simple-handler--south vue-simple-handler--east-south"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--east-north">
                <div class="vue-handler-wrapper vue-handler-wrapper--east-north vue-simple-handler-wrapper vue-simple-handler-wrapper--east vue-simple-handler-wrapper--north vue-simple-handler-wrapper--east-north vue-bounding-box__handler vue-bounding-box__handler--east-north">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--east vue-simple-handler--north vue-simple-handler--east-north"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--east">
                <div class="vue-handler-wrapper vue-handler-wrapper--east vue-simple-handler-wrapper vue-simple-handler-wrapper--east vue-bounding-box__handler vue-bounding-box__handler--east">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--east"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--west-south">
                <div class="vue-handler-wrapper vue-handler-wrapper--west-south vue-simple-handler-wrapper vue-simple-handler-wrapper--west vue-simple-handler-wrapper--south vue-simple-handler-wrapper--west-south vue-bounding-box__handler vue-bounding-box__handler--west-south">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--west vue-simple-handler--south vue-simple-handler--west-south"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--west-north">
                <div class="vue-handler-wrapper vue-handler-wrapper--west-north vue-simple-handler-wrapper vue-simple-handler-wrapper--west vue-simple-handler-wrapper--north vue-simple-handler-wrapper--west-north vue-bounding-box__handler vue-bounding-box__handler--west-north">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--west vue-simple-handler--north vue-simple-handler--west-north"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--west">
                <div class="vue-handler-wrapper vue-handler-wrapper--west vue-simple-handler-wrapper vue-simple-handler-wrapper--west vue-bounding-box__handler vue-bounding-box__handler--west">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--west"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--south">
                <div class="vue-handler-wrapper vue-handler-wrapper--south vue-simple-handler-wrapper vue-simple-handler-wrapper--south vue-bounding-box__handler vue-bounding-box__handler--south">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--south"></div>
                  </div>
                </div>
              </div>
              <div class="vue-bounding-box__handler vue-bounding-box__handler--north">
                <div class="vue-handler-wrapper vue-handler-wrapper--north vue-simple-handler-wrapper vue-simple-handler-wrapper--north vue-bounding-box__handler vue-bounding-box__handler--north">
                  <div class="vue-handler-wrapper__draggable">
                    <div class="vue-simple-handler vue-simple-handler--north"></div>
                  </div>
                </div>
              </div>
            </div>
          </div><canvas style="display: none;"></canvas><canvas style="display: none;"></canvas>
        </div>
      </div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
